### ISSUE TITLE
Analyzer: Unjustified promotion to `X & S` (promoted type variable)

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#35305]
Consider the following program:

```dart
void f<X extends num, Y extends X>(Y y) {
  if (y is int) {
    y.isEven; // Should be an error.
  }
}

main() {
  f<num, int>(42);
}
```

`dartanalyzer` version 2.2.0-dev.0.0 reports no errors for this program. The promotion to `int` (which must occur because otherwise `y.isEven` would be an error) should be justified either because `int` is a subtype of the static type of `y` at the point (which cannot be proven, and shouldn't be provable, e.g., we could have `X == Y == Null`), or it should be justified because `int` is a subtype of the bound of `Y`, in which case `y` should get the promoted type `Y & int` (but it is still not provable that `int <: X`, and shouldn't be so, and hence we shouldn't perform that promotion).

**Comment by stereotype441** [comment]
Agreed that this is a bug.  The correct behavior is for `y` not to be promoted.

Possibly related: #35306.

**Comment by MichaelRFairhurst** [comment]
I want to question whether `y` **should** _not_ be promoted.

I get that we don't want to promote this:

```
Iterable<int> xs;
if (xs is List<num>) {
  xs.first.isEven; // but this would have worked if we didn't promote
}
```

but that doesn't apply here

```
if (y is int) {
  // if we promote y to Y extends int, then we don't lose any members from Y extends num
  y.isFinite;
  // and we get the new members from int -- removing these may effectively be a breaking change
  y.isEven;
  // and we don't have any problems related to type parameter bounds
  y = 4; // still an error if we promote to Y extends int, which is good, because y could be Null
}
```

it seems like this would be an easy backwards compatible useful spec change rather than analyzer change. Something like, "type promotion of a variable of type S occurs when checked type T is such that S <: T or T is a type parameter type who's bound B is such that B <: T"

I would get, though, if the algorithm that accounts for these cases is much more complicated than that, and that this may introduce effectively a new type of subtype checking which is consistently subtly different.

**Comment by eernstg** [comment]
@MichaelRFairhurst wrote:

> I want to question whether `y` **should** _not_ be promoted.

We're currently specifying when this kind of promotion is applicable: [CL 85921](https://dart-review.googlesource.com/c/sdk/+/85921). The basic rules are that promotion to a type of the form `X & S` where `X` is a type variable and `S` is a subtype of the bound of `X` can take place at `v is S` for a local variable `v` whose declared type is `X` if `X & S` is a subtype of the type of `v` at that point.

So we won't promote anything other than a local variable (as usual) `v`, we won't promote it to such a type unless the variable has a type which is a type variable `X` (e.g., `List<X>` won't do), we won't promote to a type which isn't a subtype of the bound, and we won't promote to anything other than a subtype of what we already know.

But you could certainly argue that we should use this kind of promotion in additional situations, and this would be a good time to get it clarified.

So let's consider the examples:
```dart
Iterable<int> xs;
if (xs is List<num>) {
  xs.first.isEven; // but this would have worked if we didn't promote
}
```
In this case the type of the variable is not a type variable, so promotion to a type of the form `X & S` is not even on the table. I'm not sure how this would serve as an argument about how/whether to promote to `X & S`, but the reason why we don't promote in this situation is that `List<num>` is not a subtype of `Iterable<int>`. So maybe you're saying that we need to maintain that property? But that is indeed what we do, because we will only promote from `X` to `X & S` or from `X & T` to `X & S`, and always to a subtype of what was previously known.

Here's the other example (where I've added the context that I'm assuming you meant):
```dart
void f<X extends num, Y extends X>(Y y) {
  if (y is int) {
    // (1) if we promote y to Y extends int, then we don't lose any members from Y extends num
    y.isFinite;
    // (2) and we get the new members from int -- removing these may effectively be a breaking change
    y.isEven;
    // (3) and we don't have any problems related to type parameter bounds
    y = 4; // still an error if we promote to Y extends int, which is good, because y could be Null
  }
}
```

At (1), let's assume that we promote `y` to `Y & int`, that is, we drop the requirement that we can only promote to a type of the form `X & S` when `S` is a subtype of the bound of `X`.

It looks like we could do such a thing without ill effects&mdash;as you mention, the known members for an instance of type `Y` used to be exactly what we can get from `num`, because that's the least supertype of the type `Y` which is a class type, and from that point it will be the members from `int`, which is a subtype.

So we could consider allowing promotion from to a type `X & S` whenever `S` is a subtype of the least non-type-variable type which is a supertype of `X`, i.e., the first type we encounter as we traverse bounds which is not a type variable.

However, this violates one of the properties that we have otherwise maintained: We may now know _less_ than we used to know:

```dart
void f<X extends num, Y extends X>(Y y) {
  if (y is int) {
    // Assume that we promote `y` to have type `Y & int`. If this means that, when `y` is
    // considered, the bound of `Y` is `int`, then the following is a compile-time error:
    X x = y;
  }
}
```

Checking `dartanalyzer` (a fresh one from commit 5451ac62e26caeacd8a8c4a660ac320bdf9c84f3 as well as the released one from 2.2.0-dev.0.0) this is actually the behavior we have currently (in the analyzer, not in the vm).

So I'm not quite sure we can claim 'we don't have any problems related to type parameter bounds', as stated at (3).

One more thing around (3): If we have any assignments to `y`, we won't get a promotion. So this will make `y.isEven` an error in the first place, which doesn't seem to be the intention.

So, my conclusion: I do not think we should explore the relaxation of promotion to types of the form `X & S` as described above (just require that `S` is a subtype of the least supertype which is not a type variable), we should preserve the current rules where `S` must be a subtype of the bound of `X` (as well as the other requirements, that we haven't discussed any changes for).

**Comment by eernstg** [comment]
Thanks!
